libs/ml/include/ml/ops/max_pool.hpp

```diff
@@ -0,0 +1,188 @@
+#pragma once
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "ml/exceptions/exceptions.hpp"
+#include "ml/ops/max_pool_1d.hpp"
+#include "ml/ops/max_pool_2d.hpp"
+#include "ml/ops/ops.hpp"
+
+#include <cassert>
+#include <memory>
+#include <vector>
+
+namespace fetch {
+namespace ml {
+namespace ops {
+
+template <class T>
+class MaxPool : public Ops<T>
+{
+public:
+  using TensorType    = T;
+  using SizeType      = typename TensorType::SizeType;
+  using DataType      = typename TensorType::Type;
+  using ArrayPtrType  = std::shared_ptr<TensorType>;
+  using VecTensorType = typename Ops<T>::VecTensorType;
+  using SPType        = OpMaxPool1DSaveableParams<T>;
+  using MyType        = MaxPool<TensorType>;
+
+  MaxPool(SizeType const kernel_size, SizeType const stride_size)
+    : kernel_size_{kernel_size}
+    , stride_size_{stride_size}
+    , max_pool_1d_(kernel_size, stride_size)
```

Comment:
so in tensorflow the way this presumably gets handled is that whenever you add an op you're supposed to specify the relevant shape of the input (except the batch size which can be 'None'); so presumably they can just deduce whether it's a 1D or 2D maxpool based on the input data shape.

We could do the same thing (or just have a flag for 1D or 2D mode)? Then we wouldn't have to initialise both the 1D and 2D max pool ops on construction? In fact the MaxPool Op can just store a shared_ptr to an Ops, which can be dynamically cast to the correct type on construction, right?

In the future we may also wish to separate graph definitions and graph compilation more thoroughly, and this will require us to pre-specify input dimensions for ops anyway.